storage/spanner.go

```diff
@@ -0,0 +1,984 @@
+							return false, err
+						}
+					}
+				}
+			}
+			delete(expr.AddValues, expr.Field[index])
+		}
+		for k, v := range expr.AddValues {
+			val, ok := rowMap[k].(float64)
+			if ok {
+				m[k] = val + v
+				err = checkInifinty(m[k].(float64), expr)
+				if err != nil {
+					return false, err
+				}
+
+			} else {
+				m[k] = v
+			}
+		}
+	}
+	for i := 0; i < len(e.Attributes); i++ {
+		e.ValueMap[e.Tokens[i]] = evaluateStatementFromRowMap(e.Attributes[i], e.Cols[i], rowMap)
+	}
+
+	status, err := utils.EvaluateExpression(e)
+	if err != nil {
+		return false, err
+	}
+	return status, nil
+}
+
+func evaluateStatementFromRowMap(conditionalExpression, colName string, rowMap map[string]interface{}) interface{} {
+	if strings.HasPrefix(conditionalExpression, "attribute_not_exists") || strings.HasPrefix(conditionalExpression, "if_not_exists") {
+		if len(rowMap) == 0 {
+			return true
+		}
+		_, ok := rowMap[colName]
+		if ok {
+			return false
+		}
+		return true
+	}
+	if strings.HasPrefix(conditionalExpression, "attribute_exists") || strings.HasPrefix(conditionalExpression, "if_exists") {
+		if len(rowMap) == 0 {
+			return false
+		}
+		_, ok := rowMap[colName]
+		if ok {
+			return true
+		}
+		return false
+	}
+	return rowMap[conditionalExpression]
+}
+
+func (s Storage) performPutOperation(ctx context.Context, t *spanner.ReadWriteTransaction, table string, m map[string]interface{}) error {
+	ddl := models.TableDDL[table]
+	for k, v := range m {
+		t, ok := ddl[k]
+		if t == "BYTES(MAX)" && ok {
+			ba, err := json.Marshal(v)
+			if err != nil {
+				return errors.New("ValidationException", err)
+			}
+			m[k] = ba
+		}
+	}
+
+	mutation := spanner.InsertOrUpdateMap(table, m)
+	mutations := []*spanner.Mutation{mutation}
+	err := t.BufferWrite(mutations)
+	if e := errors.AssignError(err); e != nil {
+		return e
+	}
+	return nil
+}
+
+// SpannerBatchPut - this insert or update data in batch
+func (s Storage) SpannerBatchPut(ctx context.Context, table string, m []map[string]interface{}) error {
+	mutations := make([]*spanner.Mutation, len(m))
+	ddl := models.TableDDL[changeTableNameForSP(table)]
+	table = changeTableNameForSP(table)
+	for i := 0; i < len(m); i++ {
+		for k, v := range m[i] {
+			t, ok := ddl[k]
+			if t == "BYTES(MAX)" && ok {
+				ba, err := json.Marshal(v)
+				if err != nil {
+					return errors.New("ValidationException", err)
+				}
+				m[i][k] = ba
+			}
+		}
+		mutations[i] = spanner.InsertOrUpdateMap(table, m[i])
+	}
+	_, err := s.getSpannerClient(table).Apply(ctx, mutations)
+	if err != nil {
+		return errors.New("ResourceNotFoundException", err.Error())
+	}
+	return nil
+}
+
+// SpannerDelete - this will delete the data
+func (s Storage) SpannerDelete(ctx context.Context, table string, m map[string]interface{}, eval *models.Eval, expr *models.UpdateExpressionCondition) error {
+	_, err := s.getSpannerClient(table).ReadWriteTransaction(ctx, func(ctx context.Context, t *spanner.ReadWriteTransaction) error {
```

Comment:
This could probably also just use the `Client.Apply(...spanner.Mutation)` method in the same way as `SpannerBatchPut`, instead of defining a `ReadWriteTransaction`.

Reply:
@rosspatil, your thoughts?